test/libgit2.jl

```diff
@@ -2403,17 +2403,27 @@ mktempdir() do dir
                     payload = CredentialPayload(Nullable{AbstractCredentials}(),
                                                 Nullable{CachedCredentials}(), cfg,
                                                 allow_git_helpers=true)
-                    credential_loop($valid_cred, $url, Nullable{String}(), payload)
+                    err, auth_attempts = credential_loop($valid_cred, $url,
+                                                         Nullable{String}(), payload)
+
+                    # Triggering a garbage collection will cause the internal GitCredential
+                    # used in `authenticate_userpass` to be securely wiped but this should
+                    # not cause the payload credential to be wiped (#24731).
+                    gc()
```

Comment:
Wouldn't it be better to add a manual `finalize()` call in `authenticate_userpass`?

Reply:
I can add an explicit `securezero!` call in `authenticate_userpass` but this `gc()` will still be required to ensure that the memory has been duplicated.

Reply:
What do you mean by `gc` ensuring that memory is duplicated?

Reply:
>What do you mean by gc ensuring that memory is duplicated?

The original issue was that in `authenticate_userpass` sensitive information was being [loaded into `git_cred`](https://github.com/JuliaLang/julia/blob/f8243b7d203a28c1ac4374f883fcb6d4d9667418/base/libgit2/callbacks.jl#L188) and then [shallow copied into a `UserPasswordCredentials` credential](https://github.com/JuliaLang/julia/blob/f8243b7d203a28c1ac4374f883fcb6d4d9667418/base/libgit2/callbacks.jl#L190). Since the data was being shallow copied once the `git_cred` was garbage collected (and then finalized) the data stored within the `cred` would also be securely wiped.

The `gc()` in this test case ensures that the `git_cred` is consistently garbage collected (this was issue was previously only cropping up on 32-bit systems) which will cause the `git_cred` data to be wiped out. Since the `git_cred` information is now deep copied to 
 `cred` the secure wiping of `git_cred` should not impact the data within `cred`.

Reply:
Adding a manual `finalize()` call would cause the `git_cred` memory to be wiped sooner (definitely good idea) and would also make sure that `cred` and `git_cred` data is independent. My concern with removing the `gc()` call and only having a `finalize()` call only is that the conditions to reproduce this issue exist within the source code and not within the tests.

Potentially a future developer could remove `deepcopy` calls and the `finalize` calls and the tests would still pass as long as garbage collection hasn't happened to run.

Reply:
With all of that said having a manual `gc()` call in a test case is a strange thing to do. If we do  consider it bad practise we can go with just using `finalize()` and hope only me or stevengj modify this code ;)

Reply:
Just having it in a test case is fine.

Reply:
@stevengj Good to know. I'll merge as is, without the `gc` call, so we can get this resolved.